xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -94,38 +122,347 @@ define(["jquery"], function ($) {
+   * @param {String} filter.value Value for the new filter entry
+   */
+  Logic.prototype.filter = function (property, index, filterEntry) {
+    var self = this;
+    if (this.data.query.properties.indexOf(property) === -1) { return; }
+    // default index
+    index = index || 0;
+    if (index < 0) { return; }
+    // filter entry at current index
+    var queryFilter = this.data.query.filters[property] || [];
+    var currentEntry = queryFilter[index] || {};
+    // default filterEntry
+    filterEntry = filterEntry || {};
+    var defaultFilterEntry = {
+      property: property,
+      value: "",
+      get operator () {
+        return self.getFilterDefaultOperator(this.property);
+      },
+    };
+    filterEntry = $.extend({}, defaultFilterEntry, currentEntry, filterEntry);
+    if (filterEntry.operator === undefined) { return; }
+    // apply filter
+    // remove filter at current property and index
+    if (queryFilter[index]) {
+      queryFilter.splice(index, 1);
+    }
+    // add filter at new property and index
+    if (!this.data.query.filters[filterEntry.property]) {
+      this.data.query.filters[filterEntry.property] = [];
+    }
+    this.data.query.filters[filterEntry.property].splice(index, 0, filterEntry);
+    // dispatch events
+    this.triggerEvent("filter", {
+      property: filterEntry.property,
+      operator: filterEntry.operator,
+      value: filterEntry.value,
+      index: index,
+    });
+
+    // CALL FUNCTION TO FETCH NEW DATA HERE
+  };
+
 
-    return defer.promise();
+
+  /**
+   * Add new filter entry, shorthand of Logic.prototype.filter
+   * @param {String} property Which property to add the filter to
+   * @param {String} operator The operator of the filter. Should match the filter descriptor of the property
+   * @param {String} value Default value for the new filter entry
+   */
+  Logic.prototype.addFilter = function (property, operator, value) {
+    var index = (this.data.query.filters[property] || []).length;
+    this.filter(property, index, {
+      property: property,
+      operator: operator,
+      value: value
+    });
   };
 
 
 
+  /**
+   * Remove a filter entry in the configuration, then fetch new data
+   * @param {String} property Which property to add the filter to
+   * @param {String} index The index of the filter to remove. Undefined means last.
+   */
+  Logic.prototype.removeFilter = function (property, index) {
+    if (this.data.query.properties.indexOf(property) === -1) { return; }
+    if (!this.data.query.filters[property]) { return; }
+    // default index
+    if (index === undefined) {
+      index = this.data.query.filters[property].length - 1;
+    }
+    if (index < 0) { return; }
+    // remove filter
+    this.data.query.filters[property].splice(index, 1);
+    // dispatch events
+    var event = new CustomEvent("xwiki:livedata:removeFilter", {
+      livedata: this,
+      property: property,
+      index: this.data.query.filters[property].length - 1,
```

Comment:
I've fixed it, now the index returns the index where the filter was removed.
I also added the filter object that was removed in a `removedFilter` property, because that could be useful too.